fperms_iscore/permissions.py

```diff
@@ -56,21 +58,32 @@ def _get_perm(self, obj):
                 Q(object_id=None)
             )
         )
-        try:
-            return perms.get()
-        except Perm.DoesNotExist:
-            return None
+        return perms.order_by('-object_id').first()
 
     def has_permission(self, name, request, view, obj=None):
         """
         Return `True` if user has set permission of is superuser
         """
+        cache = self._get_cache(request).get(self.name, {})
         if not super().has_permission(name, request, view, obj):
             return False
+        elif obj is None:
+            has_general_perm = cache.get(GENERAL_CACHE_NAME)
+            if has_general_perm is None:
+                has_general_perm = request.user.perms.has_perm(self._get_perm())
+                self._add_to_cache(request, GENERAL_CACHE_NAME, has_general_perm)
+            return has_general_perm
         else:
-            try:
-                return self._get_cache(request)[self.name][(request.user.pk, obj.pk if obj else None)]
-            except KeyError:
-                 has_perm = request.user.perms.has_perm(self._get_perm(obj=obj))
-                 self._add_to_cache(request, obj, has_perm)
-                 return has_perm
+            has_general_perm = cache.get(GENERAL_CACHE_NAME)
+            if has_general_perm:
+                return has_general_perm
+
+            has_obj_perm = cache.get(str(obj.pk))
+            if has_obj_perm is None:
+                perm = self._get_perm(obj=obj)
+                has_obj_perm = request.user.perms.has_perm(perm)
+                self._add_to_cache(request, str(obj.pk), has_obj_perm)
+                if perm and not perm.object_id:
```

Comment:
if perm object has no object_id it means that was returned general permission, we can cache it too